### PR TITLE
Optimistic Relay

### PR DESCRIPTION
**Pull request by michaelneuder**
**NOTE: We have made some minor revisions to the implementation based on our learnings of running the relay on Goerli, we are closing this PR in favor of a newer one on a fresh branch again with only 2 commits: https://github.com/flashbots/mev-boost-relay/pull/285**

## 📝 Summary

*tl;dr;* This PR implements an "optimistic" version of flashbots' mev-boost-relay. The core idea is that we can improve the throughput and latency of block submissions by using asynchronous processing. Instead of validating each block that comes in immediately, we return early and add the block to the datastore before checking its validity. This (hopefully) allows builders to submit more blocks, and critically, submit blocks later in the time period of the slot (because they don't have to wait for the simulation of the block to return successfully). All blocks sent to the relay are still validated, and only a single slot at a time is being processed optimistically. In practice we expect that the first few seconds of `Slot n` will handle the `Slot n-1` optimistic blocks, which may have been received very close to the slot boundary.  Proposers still use the Proposer API to select the highest bid, but there is no longer a guarantee that this block is valid (because we may not have had time to asynchronously validate it). To submit blocks in the optimistic mode, a builder must put up a collateral that is greater than the value of the blocks they are proposing. If a proposer ends up signing an invalid block, collateral from the builder of that block will be used to refund the proposer for the missed slot. 

To aid with readability, we split the PR into two commits:

(a) functional changes --- https://github.com/flashbots/mev-boost-relay/commit/55a5eaf5651fbd46908502615b5ee365081104ba 

(b) tests --- https://github.com/flashbots/mev-boost-relay/commit/d4b9aade062a8880ac6bd23ffab81aeef05e054a

## 📚 References

This joint work from Justin Drake, AlphaMonad, and I is a revision of https://github.com/michaelneuder/mev-boost-relay/pull/2. Many thanks to Chris Hager, Alex Stokes, Mateusz Morusiewicz, and Builder0x69 for feedback and support thus far!! 

## ⛱ Motivation and Context

The changes can be described through 3 sequences:
1. Submitting optimistic blocks.
2. Validating optimistic blocks.
3. Proposing optimistic blocks.
---

### 1. Submitting optimistic blocks.
Screen Shot 2023-02-02 at 11 31 24 PM

0. Block builders submit blocks to the `Builder API` endpoint of the relay.
1. Based on the collateral and status of the builder:
a. if the builder collateral is greater than the value of the block and the builder is not `demoted`, run the block simulation optimistically.
b. otherwise if the builder is `highPrio` send the block to the `highPrio` queue of the `prio-load-balancer`.
c. else send the block to the `lowPrio` queue of the `prio-load-balancer`.
2. For non-optimistic blocks, wait for the validation result.
3. Update the builder's current bid in redis. 

> Notice that for builders with sufficient collateral, we update the bid without validating the incoming block (though we queue it for async processing). This is where the improved performance can (hopefully) be achieved. 

---
### 2. Validating optimistic blocks.
Screen Shot 2023-02-13 at 9 36 53 PM

0. The `OptimisticBlockProcessor` sends the block as `low-prio` to the `prio-load-balancer` for simulation. 
1. The block is simulated on the validating nodes, and the status is returned to the `OptimisticBlockProcessor`.
2. If the simulation failed, the builder is demoted and the details of the failure are written to the database.

> This flow handles the simulation of all the blocks that were optimistically skipped. An invalid block here results in a demotion, but not necessarily a refund for the proposer because we don't yet know whether this block was the winning bid and thus signed + proposed.

---
### 3. Proposing optimistic blocks
Screen Shot 2023-02-05 at 4 39 33 PM

0. `mev-boost` calls `getHeader` on the `Proposer API` of the relay. This is part of the `MEV-Block Block Proposal` as documented in https://docs.flashbots.net/flashbots-mev-boost/architecture-overview/block-proposal. 
1. `mev-boost` calls `getPayload` on the `Proposer API` of the relay. This triggers the publication of a `SignedBeaconBlock`. 
2. The block that was just proposed is sent to the high-prio queue on the `prio-load-balancer` for simulation.
3. The `prio-load-balancer` runs the simulation of the block against the validation nodes.
4. If the block simulation failed, an invalid block was published. We upsert a new row into the `Builder Demotions Table` with the details of the failure so the refund can be justified.

> This flow represents the process of checking if our optimistic block processing ever results in a validator submitting an invalid block. Since block builders will post collateral, this will be used to reimburse the validator in that case. Since refunds should be a relatively rare event, we plan on handling them manually.

---
### Misc notes

> We only allow optimistic processing of one slot at a time. We use a waitgroup to ensure that before we update the optimistic slot, all the previous slot blocks have been processed. This may bleed into the subsequent slot, but that is OK because we are actually shifting that processing time from the end of the previous slot which is where the timing is much more critical for winning bids. 

> At the beginning of each slot we also cache the status and collateral of each builder. We access this cache during the block submission flow to (1) avoid repeatedly reading the same status + collateral for a builder and (2) avoid race conditions where the status + collateral of a builder changes over course of a slot due to either a demotion or an internal API call. 

> Since builders may use many different public keys to submit blocks we allow all of those keys to be backed by a single collateral through the use of the a "Collateral ID". This is aimed at simplifying the process of posting collateral, but if any public key using that Collateral ID submits an invalid block, all of the public keys are demoted. 

> We introduce a database migration by adding `is_demoted, collateral_value, collateral_id` as columns to the block builder table. We also introduce one new table for the demotions. Additionally, we introduce an internal API on the path `/internal/v1/builder/collateral/{pubkey}` to provide a convenient way of updating the collateral of a builder in the DB. 

> The unit tests are part of a broader test plan for these changes that is presented in https://docs.google.com/document/d/1k_R6qJLfzpm-rA7pbkb3PF5_PJ4CZY58x6TzQ5c8fN8/edit#. We are moving into the integration and e2e testing phase of that doc, where we stand up the new version of the relay on goerli testnet. Running on goerli with a fake builder will improve our overall confidence that the changes are WAI.
---

## ✅ I have run these commands

* [x] `make lint`
* [x] `make test-race` (this gave a few errors, but i think it is likely because we are doing a lot of asynchronous processing now, not quite sure how to deal with this)
* [x] `go mod tidy`
* [x] I have seen and agree to `CONTRIBUTING.md`

### CODE REVIEW

**Comment by michaelneuder** [comment]
### Optimistic relay FAQ by Justin Drake
---

**Does an optimistic relay simulate blocks?**

_Yes, an optimistic relay simulates all blocks that could have been forwarded to a proposer. Simulation just happens away from the latency-critical path._

**Can optimistic relaying lead to mass missed slots?**

_Assuming no relay bug, the worst case is one missed slot per collateralised builder (prior to reactivation). The beacon chain is designed to handle missed slots._

**Are builders incentivised to produce invalid blocks?**

_The builder of an invalid winning block suffers a financial loss. Moreover, a single invalid block will disable optimistic relaying for the builder, yielding a latency disadvantage pending reactivation._

**What recommendations do you have for optimistic relay operators?**

_We have several recommendations for optimistic relay operators:_

  * _alerts – setup automatic alerts (e.g. email or phone call) for demotions_
  * _refunds – promptly transfer (e.g. within 24 hours) the bid value plus beacon chain penalties and missed rewards to the proposer fee recipient_
  * _max collateral – cap the maximum collateral amount per builder (e.g. to 10 ETH) to keep a level playing field_
  * _investigation – manually investigate demotions and ask builders to fix block building bugs before reactivating optimistic relaying_
  * _cool-off – impose a post-demotion cool-off period (e.g. 24 hours) before reactivating optimistic relaying_
  * _penalty – consider a fixed penalty (e.g. 0.1 ETH) per demotion, especially for repeat demotion_

File: services/api/service.go [hunk 1]
```diff
@@ -835,6 +966,55 @@ func (api *RelayAPI) handleGetPayload(w http.ResponseWriter, req *http.Request)
...
+		submitBlockReq := types.BuilderSubmitBlockRequest{
+			Signature:        payload.Signature,
+			Message:          &bidTrace.BidTrace,
+			ExecutionPayload: getPayloadResp.Data,
+		}
+
+		simErr := api.simulateBlock(blockSimOptions{
+			ctx:        context.Background(),
+			log:        log,
+			isHighPrio: true, // manually set to true for these blocks.
```
Comment (JustinDrake) [inline services/api/service.go:980]: This should probably be `isHighPrio: false`. The reason is that all blocks (including optimitic ones) will get simulated within the slot and we should prioritise the simulation of high-prio-non-optimistic blocks over optimistic blocks.

Comment (michaelneuder) [reply services/api/service.go:980]: good point! updated and changed the description and diagram to reflect this :-)

**Comment by michaelneuder** [comment]
Closing in favor of the new PR based on some implementation changes: https://github.com/flashbots/mev-boost-relay/pull/285

**Pull request closed without merging**